### ISSUE TITLE
Document where rpms should store UKI addons

### ISSUE THREAD

**Issue by esposem** [systemd/systemd#29372]
### Component

bootctl, systemd-stub

### Is your feature request related to a problem? Please describe

There will be rpms shipped by companies (like RH) that will contain addons that could be optionally installed in the ESP.
How to install it's up to the user, it could be also with a simple `cp`, but I think we all need to agree on a common location on where addons should be installed in the root fs. 
Ideally also document it somewhere (here, somewhere else?) upstream so that all others can follow a common pattern.

Currently proposals are `usr/lib/kernel/addons/` or `/usr/share/uki/addons`. Ideas?

**Comment by esposem** [comment]
You mean `/usr/lib/systemd/boot/efi/extra/`? For me whatever works, but there needs to be agreement and clear documentation. @bluca  @berrange  @vittyvk

**Comment by esposem** [comment]
@kraxel too :)

**Comment by kraxel** [comment]
Both `/usr/lib/systemd/boot/efi` and `/usr/lib/kernel` make sense to me.
Another question is whenenever we want place boths addons and sysexts in a `extra` subdir or whenever we want separate `addons` and `sysexts` subdirectories ...

**Comment by vittyvk** [comment]
Personally, I'd say that `/usr/lib/kernel` is a bit more suitable for UKI addons as these have little to do with `systemd` (yes, systemd-stub code is likely used to create them, but that's it). On the contrary, system extension images (`sysext`) **are** systemd-specific and thus `/usr/lib/systemd/boot/...` is a natural place for them. In any case this is not a big deal as long as we have a common place and not each distro picking its own location.

**Comment by poettering** [comment]
So I figure you want both global and per-kernel addons, right?

We so far put kernels and kmods into `/usr/lib/modules/$(uname -r)/`. I'd probably just build on that. Maybe the most obvious way would be for an UKI kernel `/usr/lib/modules/$(uname -r)/foobar.efi` to put addons, creds, sysexts, in `/usr/lib/modules/$(uname -r)/foobar.efi.extra.d/*`?

That way, kernel-install could just copy the EFI binary annd all such drop-ins pretty much 1:1 from that dir to the ESP.

This would then leave the question open where to put global creds to, but maybe `/usr/lib/modules/_any/extra.d/*` would be appropriate for that?

**Comment by poettering** [comment]
@bluca what's your take on this?

**Comment by esposem** [comment]
@poettering do you think the systemd-stub/ukify doc should have a reference to whatever decision is taken here?

**Comment by xnox** [comment]
In Ubuntu, we currently only use `/usr/lib/modules/` for kernel modules, as some tools like to vendor / rsync / package the whole of it into initrd.

For the signed kernel.efi we currently ship them in `/usr/lib/linux/efi`. If we need per-abi .efi like things (UKIs or addons), I would prefer to keep them out of `/usr/lib/modules` and create a tree under `/usr/lib/linux` if at all possible.

Especially since some of the addons will vendor-in modules.

**Comment by poettering** [comment]
So over at the uapi group we had a discussion about this a while back, which ultimately resulted in the last part of this to be added:

https://uapi-group.org/specifications/specs/unified_kernel_image/#distribution-built-ukis-installed-by-package-managers

I think we should build on that.

**Comment by DaanDeMeyer** [comment]
> In Ubuntu, we currently only use /usr/lib/modules/ for kernel modules, as some tools like to vendor / rsync / package the whole of it into initrd.

I think this ship has sailed really, every initrd tool that wants to be cross distribution already has to deal with the fact that Fedora ships its kernels in /usr/lib/modules/vmlinuz and that's very unlikely to change again. So inventing yet another directory just so tools can avoid having to exclude some files from /usr/lib/modules does not seem worth it to me.

**Comment by poettering** [comment]
yeah, and i thought that debian might adopt /usr/lib/modules for kernel images too, no?

**Comment by poettering** [comment]
> @poettering do you think the systemd-stub/ukify doc should have a reference to whatever decision is taken here?

I'd recommend that the path is added to https://uapi-group.org/specifications/specs/unified_kernel_image/

**Comment by xnox** [comment]
> > In Ubuntu, we currently only use /usr/lib/modules/ for kernel modules, as some tools like to vendor / rsync / package the whole of it into initrd.
> 
> I think this ship has sailed really, every initrd tool that wants to be cross distribution already has to deal with the fact that Fedora, Arch and OpenSUSE ship their kernels in /usr/lib/modules/vmlinuz and that's very unlikely to change again. So inventing yet another directory just so tools can avoid having to exclude some files from /usr/lib/modules does not seem worth it to me.

Not sure what has sailed here. So far all files in fedora/opensuse under /usr/lib/modules/ appear to be unique. and UKI is combination of them vendored together. It seems odd to put them there. But obviously do whatever makes sense on the systems there.

**Comment by xnox** [comment]
> yeah, and i thought that debian might adopt /usr/lib/modules for kernel images too, no?

Eww. That's a waste of disk space and duplication no? in Debian/Ubuntu, linux-image-* packages install /boot/vmlinuz-$(uname -r) directly. Such that one doesn't need two copies of vmlinuz in `/boot` and in `/usr/lib`.

And like other packages in Fedora I thought install stuff directly into `/boot` or `/boot/efi` as well - signed shim, and signed grub?

**Comment by xnox** [comment]
> So over at the uapi group we had a discussion about this a while back, which ultimately resulted in the last part of this to be added:
> 
> https://uapi-group.org/specifications/specs/unified_kernel_image/#distribution-built-ukis-installed-by-package-managers
> 
> I think we should build on that.

which documents a path different from where Ubuntu has been shipping UKIs for 3 years now =)

**Comment by poettering** [comment]
> Eww. That's a waste of disk space and duplication no? in Debian/Ubuntu, linux-image-* packages install /boot/vmlinuz-$(uname -r) directly. Such that one doesn't need two copies of vmlinuz in `/boot` and in `/usr/lib`.

Urks, that's not really a good idea. I am not sure what /boot/ is supposed to be, but if this is something accessible by the boot loader then it is untrusted territory, shared between parallel installs, and hence generally not where package managers should operate in. On Fedora and similar modern distros the package manager doesn't manage the ESP, but let's kernel-install do this, whose job is to copy the kernels from /usr/lib/ to the ESP, and GC the kernels from there again. 

This is particularly relevant and as kernels might have auxiliary files associated with them (addons, sysexts, creds, random seed, …) which need to be copied in too. Moreover, the file names in the ESP might contain boot counters, which need to be added in on installation into the ESP. To lock down the initrd gap modern distros even might generate a UKI from a non-UKI kernel in /usr/lib/ that is augmented with a locally generated initrd, signed locally and then dropped in the ESP.

Long story short: letting the package manager manage the ESP typically not a good idea, it's just too simple. I am pretty sure Ubuntu will sooner or later notice that too.

**Comment by bluca** [comment]
> @bluca what's your take on this?

Too soon - the Debian kernel package is being heavily refactored right now by Bastian. Stopping to ship files in /boot/ is on the radar, but it is a bit further out yet.

**Comment by keszybz** [comment]
> > In Ubuntu, we currently only use /usr/lib/modules/ for kernel modules, as some tools like to vendor / rsync / package the whole of it into initrd.
> 
> I think this ship has sailed really, every initrd tool that wants to be cross distribution already has to deal with the fact that Fedora, Arch and OpenSUSE ship their kernels in /usr/lib/modules/vmlinuz and that's very unlikely to change again. So inventing yet another directory just so tools can avoid having to exclude some files from /usr/lib/modules does not seem worth it to me.

It's even worse, because we now also have `/usr/lib/modules/vmlinuz-virt.efi`, i.e. the kernel *and* modules. So copying "everything" gets recursively worse on each invocation ;)

> And like other packages in Fedora I thought install stuff directly into `/boot` or `/boot/efi` as well - signed shim, and signed grub?

They do, and it's a freaking PITA. I deeply hope that we can convince them to stop doing that, because it just breaks things left and right as soon as one tries to depart from the one layout that the package assumes, or to use shim/grub not for the machine but to build and image, etc.

**Comment by keszybz** [comment]
> I'd say that `/usr/lib/kernel` is a bit more suitable for UKI addons as these have little to do with `systemd` (yes, systemd-stub code is likely used to create them, but that's it).

I agree. This is a nice short path. So for it's only been used for stuff related to `kernel-install`, so we can add a new subdirectory there without risk of conflict or confusion.

`/usr/lib/kernel/extra.d`

> We so far put kernels and kmods into `/usr/lib/modules/$(uname -r)/`. I'd probably just build on that. Maybe the most obvious way would be for an UKI kernel `/usr/lib/modules/$(uname -r)/foobar.efi` to put addons, creds, sysexts, in `/usr/lib/modules/$(uname -r)/foobar.efi.extra.d/*`?

+1

> This would then leave the question open where to put global creds to, but maybe `/usr/lib/modules/_any/extra.d/*` would be appropriate for that?

It's an option, but `/usr/lib/kernel/extra.d` is "nicer". Also, `/usr/lib/modules/` is the territory of the kernel packages, so I'd prefer to not use it for that.

**Comment by berrange** [comment]
> > In Ubuntu, we currently only use /usr/lib/modules/ for kernel modules, as some tools like to vendor / rsync / package the whole of it into initrd.
> 
> I think this ship has sailed really, every initrd tool that wants to be cross distribution already has to deal with the fact that Fedora, Arch and OpenSUSE ship their kernels in /usr/lib/modules/$(uname -r)/vmlinuz and that's very unlikely to change again. So inventing yet another directory just so tools can avoid having to exclude some files from /usr/lib/modules does not seem worth it to me.

Using a directory with `modules` in the name for non-module content has always been conceptually dubious. Despite it being long standing Fedora/RHEL practice, I wouldn't rule out being able to change it, if someone could specify a cross-distro standard. Fedora could just make `/usr/lib/modules/$(uname -r)` symlink to  he new locations for back-compat.

IOW, I would **not** pick `/usr/lib/modules` as the long term place for UKIs / addons too, merely because it exists.

> It's even worse, because we now also have /usr/lib/modules/vmlinuz-virt.efi, i.e. the kernel and modules. So copying "everything" gets recursively worse on each invocation ;)

The UKI hasn't been around that long, so I think there's still scope for Fedora/RHEL to change that location to something saner.

> > I'd say that `/usr/lib/kernel` is a bit more suitable for UKI addons as these have little to do with `systemd` (yes, systemd-stub code is likely used to create them, but that's it).
> 
> I agree. This is a nice short path. So for it's only been used for stuff related to `kernel-install`, so we can add a new subdirectory there without risk of conflict or confusion.
> 
> `/usr/lib/kernel/extra.d`

I think `/usr/lib/kernel` would be saner in general as the root of a hierarchy for all files related to the kernel - modules, vmlinuz, UKIs, addons, kernel-install files, etc.

**Comment by xnox** [comment]
My preference would be to be a bit more explicit and use `/usr/lib/linux`, for one it is already in use in Ubuntu (but that's minor on grand scheme of things). It matches with vendor name in ESP that systemd-boot uses. And it does make it explicit that these things are still linux kernels, rather than freebsd/hurd/etc. And thus offers clear path separation for those that do ship alternative kernels (e.g. kfreebsd). It would be best not to pick ambigious name again, just like there was with modules.

**Comment by berrange** [comment]
> My preference would be to be a bit more explicit and use `/usr/lib/linux`

Yes, your rationale makes sense

**Comment by DaanDeMeyer** [comment]
> Using a directory with modules in the name for non-module content has always been conceptually dubious. Despite it being long standing Fedora/RHEL practice, I wouldn't rule out being able to change it, if someone could specify a cross-distro standard. Fedora could just make /usr/lib/modules/$(uname -r) symlink to he new locations for back-compat.

We'd also have to convince opensuse and Arch and other distros to do the same. I'm pretty sure that if we invent a new location we'll just end up with having to deal with all three. There's an xkcd about this somewhere I think.

**Comment by keszybz** [comment]
> There's an xkcd about this somewhere I think.

There is not. We're only up to 3 locations, far from the 15 in the xkcd comic.

**Comment by mbiebl** [comment]
`/usr/lib/linux` makes a lot of sense. Given that this UKI stuff is still pretty new, I wouldn't be too concerned about backwards compatibility but rather picking a proper location going forward.

**Comment by keszybz** [comment]
OK. So I think the proposal is:

- `/usr/lib/linux/extra.d/` for generic stuff
- `/usr/lib/linux/<uname>/foobar.efi.extra.d/` for kernel `foobar.efi` in version `<uname>`

I'd be fine with that.

**Comment by esposem** [comment]
Back to this issue: it occourred to me (and @behrmann pointed that too) that maybe it doesn't really make sense to put UKIs into `/usr/lib/modules` and the addons into `/usr/lib/linux`. 
Copying what he wrote in https://github.com/uapi-group/specifications/pull/91#discussion_r1515855930 :
```
I can see the following possibilities

    * go with /usr/lib/linux/$UNAME/$UKI.efi.extra.d/
    * go with /usr/lib/modules/$UNAME/$UKI.efi.extra.d/`
    * allow both as independent directories
    * allow symlinks
```

I would personally prefer `/usr/lib/modules`, as it is already being used and we don't want to change defaults.
But as alternative, symlinks could also be a solution so all parties are happy and we don't break anyone setup.

Opinions?

**Comment by behrmann** [comment]
I think, that, since we've already written `/usr/lib/modules` for the UKIs, we should stick with `/usr/lib/modules`, since as @bluca pointed out, it doesn't make an awful lot of sense to have two different hierarchies.

**Comment by esposem** [comment]
I re-submitted the PR for the uapi-group documentation https://github.com/uapi-group/specifications/pull/91 using `/usr/lib/modules`.

**Comment by keszybz** [comment]
> I re-submitted the PR for the uapi-group documentation [uapi-group/specifications#91](https://github.com/uapi-group/specifications/pull/91) using `/usr/lib/modules`.

This WFM. I think the difference between `/usr/lib/modules` and the other alternatives has no technical basis, it's a purely bikeshedding exercise, so I think it's good that you picked something.

OTOH, I think we should have one more level of dirs, I replied in the other PR.

**Comment by esposem** [comment]
Closing this as the path has been finally decided and we merged it into the docs :partying_face:
